### PR TITLE
Issue: 42: Clean up PVC after cluster is deleted

### LINKED ISSUE #42: PVC are not deleted when ZK is deleted
When a `ZookeeperCluster` resource is deleted (i.e. `kubectl delete zk example`), PVC are not automatically deleted. PVC should have owner reference information so that they are attached to the CR lifecycle.

### PR DESCRIPTION
**Pull request by Tristan1900**
### Changelog description
This PR adds a finalizer to delete PVCs after the zk cluster is deleted. Users can decide whether to enable this finalizer by configuring it in the yaml file. The operator DOES NOT delete the PVC by default, users have to explicitly enable it.

To enable:
```
...
persistence:
  reclaimPolicy: Delete
...
```

### Purpose of the change
Fix #42 

### How to test the change
- Use this branch and start the operator
- Create a zk cluster
- Delete the zk cluster
- Verify that the PVC is deleted

Signed-off-by: wenqi <Wenqi.Mou@emc.com>

### CODE REVIEW

**Comment by Tristan1900** [comment]
@adrianmo @spiegela Could you please take a look at this PR when you have time? Thanks a lot!

File: pkg/apis/zookeeper/v1beta1/zookeepercluster_types.go [hunk 1]
```diff
@@ -130,6 +135,11 @@ func (s *ZookeeperClusterSpec) withDefaults(z *ZookeeperCluster) (changed bool)
 		}
 		changed = true
 	}
+	if z.Spec.PersistentVolumeClaimCleanup != nil {
+		changed = true
+		boolTrue := true
```
[the remark this answers is not in the document]
Comment (Tristan1900) [reply pkg/apis/zookeeper/v1beta1/zookeepercluster_types.go]: Got it, will change it.

File: pkg/apis/zookeeper/v1beta1/zookeepercluster_types.go [hunk 2]
```diff
@@ -74,6 +74,11 @@ type ZookeeperClusterSpec struct {
 	// static zookeeper configuration. If no configuration is provided required
 	// default values will be provided, and optional values will be excluded.
 	Conf ZookeeperConfig `json:"config,omitempty"`
+
+	// PersistentVolumeClaimCleanup is a zookeeper operator configuration. If it's set to true,
+	// the corresponding PVCs will be deleted by the operator when zookeeper cluster is deleted.
+	// The default value is true.
+	PersistentVolumeClaimCleanup *bool `json:"cleanUpPVC,omitempty"`
```
[the remark this answers is not in the document]
Comment (Tristan1900) [reply pkg/apis/zookeeper/v1beta1/zookeepercluster_types.go]: Yes, enum is better, will change it. Thanks for the suggestion!

File: pkg/apis/zookeeper/v1beta1/zookeepercluster_types.go [hunk 3]
```diff
@@ -130,6 +135,11 @@ func (s *ZookeeperClusterSpec) withDefaults(z *ZookeeperCluster) (changed bool)
 		}
 		changed = true
 	}
+	if z.Spec.PersistentVolumeClaimCleanup != nil {
```
[the remark this answers is not in the document]
Comment (Tristan1900) [reply pkg/apis/zookeeper/v1beta1/zookeepercluster_types.go]: Yep, you are right!

File: pkg/apis/zookeeper/v1beta1/zookeepercluster_types.go [hunk 4]
```diff
@@ -363,6 +360,43 @@ func (c *ZookeeperConfig) withDefaults() (changed bool) {
...
+	// The default value is Retain.
+	VolumeReclaimPolicy VolumeReclaimPolicy `json:"reclaimPolicy,omitempty"`
+	// PersistentVolumeClaimSpec is the spec to describe PVC for the container
+	// This field is optional. If no PVC spec, stateful containers will use
+	// emptyDir as volume.
+	PersistentVolumeClaimSpec v1.PersistentVolumeClaimSpec `json:"spec,omitempty"`
+}
+
+func (p *Persistence) withDefaults() (changed bool) {
+	if p.VolumeReclaimPolicy == "" {
```
Comment (adrianmo) [inline pkg/apis/zookeeper/v1beta1/zookeepercluster_types.go]: We should check and enforce that the value is one of the accepted ones (i.e. `Delete` or `Retain`). If the value is invalid, we could use the default `Retain` policy.

Comment (Tristan1900) [reply pkg/apis/zookeeper/v1beta1/zookeepercluster_types.go]: Right! Got it.

**Comment by Tristan1900** [comment]
Thanks @adrianmo for the review! @spiegela Could you please take another look at the change when you have time? Thanks a lot!

**Pull request merged**